### PR TITLE
PSR-4 for apps

### PR DESCRIPTION
**Pull request by nickvergessen**
Sample PR on an app: https://github.com/owncloud/announcementcenter/pull/73

@DeepDiver1975 @rullzer @MorrisJobke 

The following directories remain unchanged, since they should not contain namespaces classes and have hardcoded usage. Otherwise we would always need to check two directories:
- appinfo/
- css/
- img/
- js/
- l10n/
- templates/

### CODE REVIEW

File: lib/private/Route/Router.php [hunk 1]
```diff
@@ -351,9 +351,12 @@ private function setupRoutes($routes, $appName) {
 			$appNameSpace = App::buildAppNamespace($appName);
 
 			$applicationClassName = $appNameSpace . '\\AppInfo\\Application';
+			$applicationClassNamePsr4 = $appNameSpace . '\\Application';
```
Comment (nickvergessen) [inline lib/private/Route/Router.php]: This is required, because appinfo/ remains lowercase and therefor the class can not be PSR-4 loaded

**Comment by rullzer** [comment]
Awesome stuff!

Would it make sense to have all the php files in a src dir? Just so we don't mix CamelCase and lower case in the root app folder?

**Comment by nickvergessen** [comment]
Done, also solves the hacky problemwith duplicated appinfo/.

File: lib/private/legacy/app.php [hunk 2]
```diff
@@ -155,6 +158,22 @@ public static function loadApp($app, $checkUpgrade = true) {
 	}
 
 	/**
+	 * @param string $app
+	 * @param string $path
+	 */
+	protected static function registerAutoloading($app, $path) {
+		// Register on PSR-4 composer autoloader
+		$appNamespace = \OC\AppFramework\App::buildAppNamespace($app);
+		\OC::$composerAutoloader->addPsr4($appNamespace . '\\', $path . '/src/', true);
```
Comment (MorrisJobke) [inline lib/private/legacy/app.php]: so we move everything in a subfolder called `src`? Fine with me, but want to point this out.

Comment (MorrisJobke) [inline lib/private/legacy/app.php]: Also good to document this ;)

**Comment by rullzer** [comment]
Looks good and looks clean!
I'll tests later more!

But awesome! :balloon:

File: lib/private/legacy/app.php (hunk 2 above)
Comment (MorrisJobke) [inline lib/private/legacy/app.php]: I like it 👍

**Comment by DeepDiver1975** [comment]
> Would it make sense to have all the php files in a src dir? Just so we don't mix CamelCase and lower case in the root app folder?

Originally we had all php files living in the subfolder lib - this is also reflected in the old class loader.
Over the past years this was a bit relaxed because it simply worked to load classes from various folders.
We should revive this from my pov - all php file should live in one place

File: lib/private/legacy/app.php (hunk 2 above)
Comment (DeepDiver1975) [inline lib/private/legacy/app.php]: as explained below - let's name it lib

**Comment by MorrisJobke** [comment]
> We should revive this from my pov - all php file should live in one place

It will with this PR live in `src`. Only there the PSR-4 autoloader will work :)

**Comment by DeepDiver1975** [comment]
> It will with this PR live in src. Only there the PSR-4 autoloader will work :)

And I say: let's name it `lib` because this is where all php files should live anyhow

File: lib/private/legacy/app.php (hunk 2 above)
Comment (nickvergessen) [inline lib/private/legacy/app.php]: Yes, I will drop an explanation mail as a response on @rullzer s mail to the mailing list once this is merged.

**Comment by nickvergessen** [comment]
Switched to lib/ and adjusted the app as well

**Comment by rullzer** [comment]
Tested (with the apps @nickvergessen already moved) and works like a charm!

I agree we should add this to the docs and send to the ML. But very nice and cool stuff! :+1:

**Comment by MorrisJobke** [comment]
Tested and works like a charm with old and new apps 👍

**Pull request merged**